File: mobile/lib/timeline/timeline.dart
import '../api/api_client.dart';
import '../render/render_spec.dart';

const todoAssetIcon = '📋';
const eventAssetIcon = '📅';
const contactAssetIcon = '👤';
const notesAssetIcon = '✍️';

/// Icon + label + accent for a skill / derived kind.
class SkillMeta {
  final String icon;
  final String label;
  final String accentColor; // blue|amber|green|red|purple|gray|neutral
  /// Present for registered user skills (`/api/skills` row id) — enables the
  /// category-detail delete control. Null for built-in / first-class kinds.
  final String? userSkillId;

  /// Active-set flag: enabled skills show in the library grid + the agent
  /// routes to them. Disabled ones live only in the 技能管理页.
  final bool enabled;
  const SkillMeta(
    this.icon,
    this.label, [
    this.accentColor = 'gray',
    this.userSkillId,
    this.enabled = true,
  ]);
}

/// One unified timeline entry (asset / event / contact / input_turn / file),
/// matching the backend /api/timeline shape.
class TimelineItem {
  final String kind;
  final String id;
  final DateTime effectiveAt; // local time
  final String title;
  final String subtitle;
  final String? skillName;
  final String? sessionId;

  // Event-only: end time / all-day flag / location — needed by the day view's
  // hour grid. Null/false for non-events.
  final DateTime? endAt;
  final bool allDay;
  final String? location;
  final String? eventId;
  final String? contactId;

  /// Raw payload (asset / contact) — lets the day view render a full SkillCard.
  final Map<String, dynamic> payload;

  /// For flash (input_turn) captures: {skill_name|"event"|"contact": count}.
  final Map<String, int> derived;

  /// §4.5.0a 落段:user 只说了模糊时段时填(凌晨/上午/中午/下午/晚上),否则 ''。
  final String period;

  /// True when the user stated a clock time (asset.occurred_at set → effectiveAt
  /// is that precise moment). Events with a start time count via `kind`.
  final bool hasClockTime;

  /// True only when the user actually scheduled this item. Unlike
  /// [effectiveAt], this never becomes true merely because the backend uses the
  /// capture time as a sorting fallback for an otherwise time-less todo.
  final bool hasScheduledTime;

  /// §8 生活领域(工作/学习/健康/运动/社交/娱乐/生活/灵感),否则 ''。流/月卡片的领域 tag。
  final String domain;

  TimelineItem({
    required this.kind,
    required this.id,
    required this.effectiveAt,
    required this.title,
    required this.subtitle,
    required this.skillName,
    required this.sessionId,
    required this.derived,
    this.endAt,
    this.allDay = false,
    this.location,
    this.eventId,
    this.contactId,
    this.payload = const {},
    this.period = '',
    this.hasClockTime = false,
    this.hasScheduledTime = false,
    this.domain = '',
  });

  factory TimelineItem.fromJson(Map<String, dynamic> j) {
    final kind = j['kind'] as String? ?? 'asset';
    final ea =
        DateTime.tryParse(j['effective_at'] as String? ?? '')?.toLocal() ??
        DateTime.now();
    final rawDerived =
        (j['derived'] as Map?)?.cast<String, dynamic>() ?? const {};
    final rawPayload =
        (j['payload'] as Map?)?.cast<String, dynamic>() ?? const {};
    final payload = kind == 'event'
        ? <String, dynamic>{
            'event_id': j['event_id'],
            'title': j['title'],
            'start_at': j['start_at'] ?? j['effective_at'],
            'end_at': j['end_at'],
            'all_day': j['all_day'],
            'location': j['location'],
            'description': j['description'],
            'attendees': j['attendees'],
            ...rawPayload,
          }
        : rawPayload;
    return TimelineItem(
      kind: kind,
      id: j['id'] as String? ?? '',
      effectiveAt: ea,
      title: j['title'] as String? ?? '',
      subtitle: kind == 'event'
          ? eventCardSummary(payload)
          : j['subtitle'] as String? ?? '',
      skillName: j['skill_name'] as String?,
      sessionId: j['session_id'] as String?,
      endAt: DateTime.tryParse(j['end_at'] as String? ?? '')?.toLocal(),
      allDay: j['all_day'] == true || j['all_day'] == 1,
      location: j['location'] as String?,
      eventId: j['event_id'] as String?,
      contactId: j['contact_id'] as String?,
      payload: payload,
      derived: {
        for (final e in rawDerived.entries)
          if (e.value is num) e.key: (e.value as num).toInt(),
      },
      period: j['period'] as String? ?? '',
      hasClockTime: j['has_clock_time'] == true,
      hasScheduledTime: j['has_scheduled_time'] == true,
      domain: j['domain'] as String? ?? '',
    );
  }
}

const _builtin = <String, SkillMeta>{
  'todo': SkillMeta(todoAssetIcon, '待办', 'blue'),
  'event': SkillMeta(eventAssetIcon, '日程', 'purple'),
  'contact': SkillMeta(contactAssetIcon, '名片', 'neutral'),
  'notes': SkillMeta(notesAssetIcon, '随记', 'amber'),
  'expense': SkillMeta('💳', '记账', 'green'),
  'external_ref': SkillMeta('🔗', '外部', 'purple'),
};

/// Built-in glyphs the client pins regardless of the server's render_spec — the
/// seed mirrors these, but the client owns the canonical look. 待办 must read as
/// "to-do" (📋), not "done" (✅). Custom user skills are unaffected.
const _pinnedIcons = <String, String>{'todo': todoAssetIcon};

String _canonicalSkillKey(String key) => switch (key.trim().toLowerCase()) {
  'calendar' => 'event',
  'note' || 'idea' || 'misc' => 'notes',
  final normalized => normalized,
};

/// Resolve a skill / derived key to its icon + label. Custom skills live only
/// in the registry, so look there first (mirrors the web derivedMeta fix).
SkillMeta resolveMeta(String key, Map<String, SkillMeta> registry) {
  final normalized = key.trim().toLowerCase();
  final registered = registry[normalized] ?? registry[key];
  final canonical = _canonicalSkillKey(normalized);
  final m = registered ?? _builtin[canonical] ?? SkillMeta('•', key);
  final pin = _pinnedIcons[canonical];
  return pin == null
      ? m
      : SkillMeta(pin, m.label, m.accentColor, m.userSkillId, m.enabled);
}

/// Resolve the canonical identity for an Asset-like timeline entry.
///
/// Flash/input-turn rows are a separate presentation and should keep their
/// lightning icon instead of passing through this resolver.
SkillMeta resolveTimelineItemMeta(
  TimelineItem item,
  Map<String, SkillMeta> registry,
) {
  if (item.kind == 'event') return resolveMeta('event', registry);
  if (item.kind == 'contact') return resolveMeta('contact', registry);
  final skillName = item.skillName?.trim();
  if (skillName != null && skillName.isNotEmpty) {
    return resolveMeta(skillName, registry);
  }
  return const SkillMeta('•', '记录');
}

Future<List<TimelineItem>> fetchTimeline(
  ApiClient api, {
  bool coreRecordsOnly = false,
}) async {
  if (coreRecordsOnly) return _fetchCoreRecordTimeline(api);
  try {
    final res = await api.getJson('/api/timeline');
    final items = (res is Map ? res['items'] : null) as List? ?? const [];
    return items
        .whereType<Map>()
        .map((e) => TimelineItem.fromJson(e.cast<String, dynamic>()))
        // The 文件 entity was removed from the app — never surface file captures.
        .where((it) => it.kind != 'file')
        .toList();
  } on ApiException catch (error) {
    if (error.statusCode != 404) rethrow;
    return _fetchCoreRecordTimeline(api);
  }
}

Future<List<TimelineItem>> _fetchCoreRecordTimeline(ApiClient api) async {
  final responses = await Future.wait([
    api.getJson('/api/user-skills'),
    api.getJson('/api/assets', query: const {'limit': 100}),
    api.getJson('/api/events', query: const {'limit': 100}),
    api.getJson('/api/flash/recordings', query: const {'limit': 200}),
  ]);
  final skillsById = <String, ({String name, String domain})>{};
  for (final raw in _coreList(responses[0], 'skills').whereType<Map>()) {
    final skill = raw.cast<String, dynamic>();
    final id = skill['id']?.toString();
    final name = skill['machine_name']?.toString();
    if (id == null || name == null || name.isEmpty) continue;
    skillsById[id] = (name: name, domain: skill['domain']?.toString() ?? '');
  }

  final items = <TimelineItem>[];
  for (final raw in _coreList(responses[2], 'events').whereType<Map>()) {
    final event = raw.cast<String, dynamic>();
    if (event['status'] == 'cancelled') continue;
    final start = DateTime.tryParse(
      event['start_at']?.toString() ?? '',
    )?.toLocal();
    if (start == null) continue;
    final id = event['id']?.toString() ?? '';
    items.add(
      TimelineItem(
        kind: 'event',
        id: id,
        effectiveAt: start,
        title: _coreTitle(event, '事件'),
        subtitle: '',
        skillName: null,
        sessionId: null,
        derived: const {},
        endAt: DateTime.tryParse(event['end_at']?.toString() ?? '')?.toLocal(),
        allDay: event['all_day'] == true,
        location: event['location']?.toString(),
        eventId: id,
        payload: event,
      ),
    );
  }

  for (final raw in _coreList(responses[1], 'assets').whereType<Map>()) {
    final asset = raw.cast<String, dynamic>();
    final skill = skillsById[asset['user_skill_id']?.toString()];
    if (skill == null) continue;
    final payload =
        (asset['payload'] as Map?)?.cast<String, dynamic>() ?? const {};
    final due = payload['due_date']?.toString();
    final explicit = _firstCoreDate([
      asset['effective_at'],
      payload['occurred_at'],
    ]);
    final semantic = switch (skill.name) {
      'todo' => _firstCoreDate([due]),
      'expense' => _firstCoreDate([payload['at'], payload['date']]),
      _ => null,
    };
    final effective =
        explicit ?? semantic ?? _firstCoreDate([asset['created_at']]);
    if (effective == null) continue;
    final hasExplicitClock = explicit != null;
    items.add(
      TimelineItem(
        kind: 'asset',
        id: asset['id']?.toString() ?? '',
        effectiveAt: effective,
        title: _coreTitle(payload, skill.name),
        subtitle:
            payload['note']?.toString() ??
            payload['description']?.toString() ??
            '',
        skillName: skill.name,
        sessionId: null,
        derived: const {},
        payload: payload,
        period: payload['period']?.toString() ?? '',
        hasClockTime: hasExplicitClock || due?.contains('T') == true,
        hasScheduledTime:
            skill.name == 'todo' &&
            (hasExplicitClock || due?.contains('T') == true),
        domain: skill.domain,
      ),
    );
  }

  for (final raw in _coreList(responses[3], 'recordings').whereType<Map>()) {
    final recording = raw.cast<String, dynamic>();
    final effective = _firstCoreDate([
      recording['captured_at'],
      recording['created_at'],
    ]);
    if (effective == null) continue;
    final title = recording['title']?.toString().trim() ?? '';
    items.add(
      TimelineItem(
        kind: 'input_turn',
        id: recording['id']?.toString() ?? '',
        effectiveAt: effective,
        title: title.isEmpty ? '闪念' : title,
        subtitle: '',
        skillName: null,
        sessionId: recording['session_date']?.toString(),
        derived: const {},
        payload: {
          'process_status': recording['process_status'],
          'session_date': recording['session_date'],
        },
      ),
    );
  }
  items.sort((a, b) => a.effectiveAt.compareTo(b.effectiveAt));
  return items;
}

DateTime? _firstCoreDate(Iterable<dynamic> values) {
  for (final value in values) {
    final raw = value?.toString().trim() ?? '';
    if (raw.isEmpty) continue;
    final parsed = DateTime.tryParse(raw)?.toLocal();
    if (parsed != null) return parsed;
  }
  return null;
}

List _coreList(dynamic response, String key) => switch (response) {
  List value => value,
  Map value => value[key] as List? ?? const [],
  _ => const [],
};

String _coreTitle(Map<String, dynamic> value, String fallback) {
  final candidate =
      value['title'] ?? value['content'] ?? value['name'] ?? value['amount'];
  final title = candidate?.toString().trim() ?? '';
  return title.isEmpty ? fallback : title;
}

/// name → {icon, label} from /api/skills (render_spec.icon + display_name).
Future<Map<String, SkillMeta>> fetchSkills(
  ApiClient api, {
  bool coreRecordsOnly = false,
}) async {
  dynamic res;
  if (coreRecordsOnly) {
    res = await api.getJson('/api/user-skills');
  } else {
    try {
      res = await api.getJson('/api/skills');
    } on ApiException catch (error) {
      if (error.statusCode != 404) rethrow;
      res = await api.getJson('/api/user-skills');
    }
  }
  final skills = switch (res) {
    List value => value,
    Map value => value['skills'] as List? ?? const [],
    _ => const [],
  };
  final out = <String, SkillMeta>{};
  for (final s in skills.whereType<Map>()) {
    final name = (s['name'] ?? s['machine_name']) as String?;
    if (name == null) continue;
    final rs = (s['render_spec'] as Map?)?.cast<String, dynamic>();
    final coreSpec = coreRecordsOnly
        ? coreRecordRenderSpec(name, s['schema'])
        : null;
    out[name] = SkillMeta(
      // Pin built-in glyphs (待办 → 📋) even for surfaces that read the meta map
      // directly (e.g. 资产库 container tiles), not just via resolveMeta.
      _pinnedIcons[name] ?? (rs?['icon'] as String? ?? coreSpec?.icon ?? '•'),
      s['display_name'] as String? ?? name,
      rs?['accent_color'] as String? ?? coreSpec?.accentColor ?? 'gray',
      (s['user_skill_id'] ?? s['id']) as String?,
      (s['enabled'] as int? ?? 1) != 0,
    );
  }
  return out;
}

/// Group items into day buckets, newest day first; items within a day ascend.
List<MapEntry<DateTime, List<TimelineItem>>> groupByDay(
  List<TimelineItem> items,
) {
  final byDay = <DateTime, List<TimelineItem>>{};
  for (final it in items) {
    final d = DateTime(
      it.effectiveAt.year,
      it.effectiveAt.month,
      it.effectiveAt.day,
    );
    byDay.putIfAbsent(d, () => []).add(it);
  }
  final days = byDay.keys.toList()..sort((a, b) => b.compareTo(a));
  return [
    for (final d in days)
      MapEntry(
        d,
        byDay[d]!..sort((a, b) => a.effectiveAt.compareTo(b.effectiveAt)),
      ),
  ];
}

File: backend/core/timeline.py
"""
Timeline assembly — Phase B v1.4.x.

Note on rendering: this module is the source of the `title` / `subtitle`
strings shown in calendar bullets + the "/api/timeline" feed. Field values
are passed through `_format_value` so an ISO datetime in the primary slot
reads as "5月30日 08:00" instead of "2026-05-30T08:00:00+08:00". Mirrors
the auto-format in frontend/src/lib/format.ts applyFormat().

Powers the「全部」filter tab in CalendarPage's Schedule view (per design
DESIGN.md §5.2 + Phase B §八). NOT a standalone page — just the data source
for that one tab.

Two concepts kept distinct:
- `created_at`     —— when the row was written to DB (audit / debug)
- `effective_at`   —— when the entity is meaningful on a user-facing timeline
                      (driven by content semantics, not row-write time)

Per-kind effective_at rule (table also lives in phase-b doc §三):

  event             →  start_at
  todo asset        →  payload.due_date  | else created_at
  expense asset     →  payload.date       | else created_at
  idea/notes/misc   →  created_at
  contact asset     →  created_at
  input_turn        →  created_at  (flash: capture moment; async ASR: completion)
  file              →  created_at  (upload moment)

The 「全部」 tab shows every kind interleaved by effective_at. Concrete-kind
tabs (event / todo / expense / ...) typically use their own endpoints with
richer per-type data (event tab needs end_at + attendees; todo tab needs
status groupings; etc.) — assemble_timeline is the unified-merge code path.
"""
from datetime import datetime, timedelta, timezone

# The app's canonical user timezone (Asia/Shanghai). Naive datetime strings on the
# timeline (LLM-emitted due_date / expense date, or manual entries) are Beijing
# WALL-CLOCK times — assuming UTC would shift them −8h onto the wrong day (e.g.
# a todo due "今天 17:00" → 17:00 UTC → 01:00 次日 Beijing).
_BEIJING = timezone(timedelta(hours=8))
from typing import Optional, Any

# Tz-aware sentinel for sort fallback (datetime.min is naive — incompatible
# with offset-aware values parsed from ISO8601)
_EPOCH_MIN = datetime(1, 1, 1, tzinfo=timezone.utc)

from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from db.models import (
    Asset, UserSkill, GlobalSkill, Event, InputTurn, File, Contact,
)


# ── effective_at per kind ─────────────────────────────────────────────────────

def _parse_iso(s: Any) -> Optional[datetime]:
    """
    Parse an ISO8601 string into a tz-aware datetime; return None on failure.
    If the string lacks a timezone offset, assume **Beijing (+08:00)** — naive
    times in this app are Beijing wall-clock (LLM-emitted due_date / expense date,
    or manual entry). Assuming UTC mis-dates them by 8h onto the wrong day.
    (A real `datetime` object reaching here is an ORM column = already UTC-aware.)
    """
    if isinstance(s, datetime):
        # Already a datetime (ORM column) — coerce to aware-UTC if somehow naive.
        return s if s.tzinfo else s.replace(tzinfo=timezone.utc)
    if not s or not isinstance(s, str):
        return None
    try:
        dt = datetime.fromisoformat(s.replace("Z", "+00:00"))
    except (ValueError, TypeError):
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=_BEIJING)   # naive string = Beijing wall-clock
    return dt


def effective_at_for_asset(asset: Asset, skill_name: str,
                           render_spec: Optional[dict] = None) -> datetime:
    """Compute effective_at for an asset based on its skill type + payload.

    Per-skill time anchor (发生型 skill): a skill may declare
    `render_spec.timeline_anchor = "<payload field>"` — the field that holds
    when the thing *happened* (a 球赛 played on 6/5, a workout done yesterday).
    When present and parseable it wins, so a match recorded today lands on its
    play date in 流 instead of today. Pure record skills (notes/灵感) declare no
    anchor → created_at, unchanged. todo/expense keep their built-in anchors.
    """
    payload = asset.payload or {}
    # §4.5.0a: user stated a clock time → occurred_at is the precise moment and
    # wins over every payload-derived anchor (it's exactly "when it happened").
    if getattr(asset, "occurred_at", None):
        return asset.occurred_at
    rs = render_spec if isinstance(render_spec, dict) else {}
    anchor = rs.get("timeline_anchor")
    if anchor:
        dt = _parse_iso(payload.get(anchor))
        if dt:
            return dt
    if skill_name == "todo":
        return _parse_iso(payload.get("due_date")) or asset.created_at
    if skill_name == "expense":
        # Legacy compatibility: older expense prompts wrote payload.at, including
        # fuzzy period canonical clocks. New writes use Asset.occurred_at/period
        # instead; keep reading `at` only so historical rows keep their order.
        return (_parse_iso(payload.get("at"))
                or _parse_iso(payload.get("date"))
                or asset.created_at)
    # idea / notes / misc / contact / (no-anchor custom) — created_at by default
    return asset.created_at


def effective_at_for_event(event: Event) -> datetime:
    return event.start_at


def _todo_due_has_clock(payload: dict) -> bool:
    """Whether a todo deadline contains a user-facing clock component.

    ``effective_at`` may fall back to ``created_at`` for sorting, so its hour is
    not evidence that the user scheduled the todo. A date-only deadline remains
    unscheduled; any ISO datetime (including midnight) is an explicit time.
    """
    due = payload.get("due_date")
    return isinstance(due, str) and "T" in due


def effective_at_for_input_turn(turn: InputTurn) -> datetime:
    """
    For flash: capture moment (= created_at when ASR is sync/inline).
    For async ASR (meeting): row created when ASR completed (= created_at also).
    Same field in both cases.
    """
    return turn.created_at


def effective_at_for_file(file: File) -> datetime:
    return file.created_at


# ── TimelineItem shape ────────────────────────────────────────────────────────
# Plain dicts (not a class) for direct JSON serialization. Frontend renders
# based on `kind`. Fields not relevant to a kind are simply omitted.
#
# Common keys:
#   kind:          "asset" | "event" | "input_turn" | "file"
#   id:            uuid
#   effective_at:  ISO8601 + TZ
#   created_at:    ISO8601 + TZ (for stable tie-break ordering)
#   title:         display title
#   subtitle:      optional secondary text
#   session_id:    if applicable
#
# Kind-specific:
#   asset:        skill_name, payload, source_input_turn_id
#   event:        event_id, end_at, location, attendees_count, files_count
#   input_turn:   source (voice/typed/imported), text_snippet, derived_count
#   file:         source_tag, file_type, asr_status


def _iso(dt: Optional[datetime]) -> Optional[str]:
    return dt.isoformat() if dt else None


import re

_ISO_DT_RE   = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}(:\d{2})?(\.\d+)?(Z|[+-]\d{2}:?\d{2})?$")
_ISO_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")

def _format_value(raw) -> str:
    """
    Stringify a payload value for display in calendar bullets.

    Auto-formats anything that looks like an ISO timestamp into 「月日 HH:MM」
    so the user doesn't see "2026-05-30T08:00:00+08:00" on the timeline.
    Other types pass through unchanged.
    """
    if raw is None:
        return ""
    s = str(raw)
    if _ISO_DT_RE.match(s):
        dt = _parse_iso(s)
        if dt:
            return f"{dt.month}月{dt.day}日 {dt.hour:02d}:{dt.minute:02d}"
    if _ISO_DATE_RE.match(s):
        try:
            y, m, d = s.split("-")
            return f"{int(m)}月{int(d)}日"
        except ValueError:
            return s
    return s


# Units were dropped per May audit (Option B). The title / subtitle are
# just the formatted value — users embed units in the value themselves
# when relevant ("150 毫升", "5 km"). Keeps multi-modal skills sane.


def _asset_item(asset: Asset, skill_name: str, render_spec: Optional[dict] = None, display_name: Optional[str] = None) -> dict:
    p = asset.payload or {}
    # Title: prefer the skill's render_spec.primary_field (matches how the card
    # renders), then common title-ish fields, then the skill's display_name.
    # Never fall back to "[skill_name]" — AI-created skills with custom payloads
    # (e.g. 跑步记录 {distance, pace}) used to surface an ugly "[running]".
    #
    # Measurement skills (跑步 distance=5, 喝水 ml=200) used to surface as
    # the bare number "5". Bundle C added primary_label / primary_unit to
    # render_spec for cards; apply the same decoration here so the calendar
    # bullet reads "距离 5 km" instead of "5". Single source of rendering
    # rule for the timeline: <label?> <value> <unit?>.
    rs = render_spec if isinstance(render_spec, dict) else {}
    pf = rs.get("primary_field")
    pf_val = p.get(pf) if pf else None
    primary_str: Optional[str] = _format_value(pf_val) if pf_val not in (None, "") else None
    title = (
        primary_str or
        p.get("content") or p.get("title") or p.get("name") or
        (f"¥{p.get('amount')}" if p.get("amount") else None) or
        display_name or skill_name
    )
    sf = rs.get("secondary_field")
    sf_val = p.get(sf) if sf else None
    if sf_val not in (None, ""):
        subtitle = _format_value(sf_val)
    else:
        subtitle = p.get("description") or p.get("merchant") or ""
    return {
        "kind":                 "asset",
        "id":                   str(asset.id),
        "effective_at":         _iso(effective_at_for_asset(asset, skill_name, render_spec)),
        "created_at":           _iso(asset.created_at),
        # §4.5.0a 落段信号:period = user 只说了模糊时段; has_clock_time = user 说了
        # 钟点(occurred_at 已 set,effective_at 即该精确时刻)。两者皆无 = 捕捉兜底。
        "period":               getattr(asset, "period", None) or "",
        "has_clock_time":       getattr(asset, "occurred_at", None) is not None,
        # Distinct from effective_at: a no-deadline todo sorts by created_at but
        # must still render in 待安排, never at its capture clock time.
        "has_scheduled_time":   (
            getattr(asset, "occurred_at", None) is not None
            or (skill_name == "todo" and _todo_due_has_clock(p))
        ),
        # §8 生活领域(工作/学习/健康/运动/社交/娱乐/生活/灵感)— drives the 流/月 卡片领域 tag.
        "domain":               getattr(asset, "domain", None) or "",
        "title":                str(title)[:120],
        "subtitle":             str(subtitle)[:120],
        "skill_name":           skill_name,
        "session_id":           str(asset.session_id) if asset.session_id else None,
        "source_input_turn_id": str(asset.source_input_turn_id) if asset.source_input_turn_id else None,
        "payload":              p,
    }


def _event_item(event: Event) -> dict:
    return {
        "kind":                 "event",
        "id":                   str(event.id),
        "effective_at":         _iso(effective_at_for_event(event)),
        "created_at":           _iso(event.created_at),
        "title":                event.title,
        "subtitle":             event.location or "",
        "event_id":             str(event.id),
        "end_at":               _iso(event.end_at),
        "location":             event.location,
        "all_day":              bool(event.all_day),
        "source_input_turn_id": str(event.source_input_turn_id) if event.source_input_turn_id else None,
    }


def _input_turn_item(turn: InputTurn, derived: Optional[dict] = None) -> dict:
    text = turn.text or ""
    derived = derived or {}
    return {
        "kind":          "input_turn",
        "id":            str(turn.id),
        "effective_at":  _iso(effective_at_for_input_turn(turn)),
        "created_at":    _iso(turn.created_at),
        "title":         text[:80] + ("…" if len(text) > 80 else ""),
        "subtitle":      "",
        "source":        turn.source,
        "session_id":    str(turn.session_id) if turn.session_id else None,
        "file_id":       str(turn.file_id) if turn.file_id else None,
        # What this capture produced, keyed by skill_name | "event" | "contact".
        # Powers the timeline ⚡ summary ("待办 ×2 · 联系人 ×1").
        "derived":       derived,
        "derived_total": sum(derived.values()),
    }


def _contact_item(contact: Contact) -> dict:
    """Timeline item for a contact (名片). Contacts are a first-class entity in
    their own table — they were missing from the timeline entirely, so a newly
    added 名片 never showed in 流 / 月 views. effective_at = created_at (a
    contact isn't scheduled; it's "when I added this person")."""
    role    = (contact.title or "").strip()      # job title
    company = (contact.company or "").strip()
    subtitle = " · ".join(x for x in (role, company) if x)
    return {
        "kind":         "contact",
        "id":           str(contact.id),
        "contact_id":   str(contact.id),
        "effective_at": _iso(contact.created_at),
        "created_at":   _iso(contact.created_at),
        "title":        (contact.name or "联系人")[:120],
        "subtitle":     subtitle[:120],
        # skill_name lets the frontend's subKindOf / DayDetailSheet grouping
        # bucket it as 名片 (they switch on skill_name === "contact").
        "skill_name":   "contact",
        "payload": {
            "name":    contact.name,
            "phone":   contact.phone,
            "company": company,
            "title":   role,
            "email":   contact.email,
            "notes":   contact.notes or [],
        },
    }


def _file_item(file: File) -> dict:
    kind_label = "🎙 闪念录音" if file.source_tag == "flash" else (
                 "📁 会议录音" if file.source_tag == "meeting" else "📎 文件")
    return {
        "kind":         "file",
        "id":           str(file.id),
        "effective_at": _iso(effective_at_for_file(file)),
        "created_at":   _iso(file.created_at),
        "title":        kind_label,
        "subtitle":     f"{file.file_type or ''} · {file.duration_sec or 0}s",
        "file_id":      str(file.id),
        "source_tag":   file.source_tag,
        "asr_status":   file.asr_status,
    }


# ── derived breakdown (flash capture → what it produced) ──────────────────────

async def _derived_breakdown(db: AsyncSession, user_id: str, turn_ids: list) -> dict:
    """Count what each flash input_turn produced — assets (by skill name),
    events, contacts — keyed by source_input_turn_id.

    Returns {turn_id: {<key>: count}} where key ∈ skill_name | "event" |
    "contact". Three grouped queries; empty input → {}.
    """
    if not turn_ids:
        return {}
    out: dict = {}

    def _bump(tid, key, cnt):
        if tid is None:
            return
        k = str(tid)
        out.setdefault(k, {})
        out[k][key] = out[k].get(key, 0) + int(cnt)

    # assets grouped by (turn, skill_name)
    rows = (await db.execute(
        select(Asset.source_input_turn_id, GlobalSkill.name, func.count())
        .join(UserSkill, Asset.user_skill_id == UserSkill.id)
        .join(GlobalSkill, UserSkill.skill_id == GlobalSkill.id)
        .where(Asset.user_id == user_id, Asset.source_input_turn_id.in_(turn_ids))
        .group_by(Asset.source_input_turn_id, GlobalSkill.name)
    )).all()
    for tid, skill, cnt in rows:
        _bump(tid, skill, cnt)

    # events grouped by turn
    rows = (await db.execute(
        select(Event.source_input_turn_id, func.count())
        .where(Event.user_id == user_id, Event.source_input_turn_id.in_(turn_ids))
        .group_by(Event.source_input_turn_id)
    )).all()
    for tid, cnt in rows:
        _bump(tid, "event", cnt)

    # contacts grouped by turn (provenance column added in migration 0003)
    rows = (await db.execute(
        select(Contact.source_input_turn_id, func.count())
        .where(Contact.user_id == user_id, Contact.source_input_turn_id.in_(turn_ids))
        .group_by(Contact.source_input_turn_id)
    )).all()
    for tid, cnt in rows:
        _bump(tid, "contact", cnt)

    return out


# ── Public: assemble_timeline ─────────────────────────────────────────────────

async def assemble_timeline(
    db: AsyncSession,
    user_id: str,
    from_dt: Optional[datetime] = None,
    to_dt:   Optional[datetime] = None,
    kinds:   Optional[set] = None,        # subset of {"asset", "event", "contact", "input_turn", "file"}
    skill_names: Optional[set] = None,    # restrict asset kind to specific skills
    limit: int = 500,
) -> list:
    """
    Assemble a unified, time-sorted list of TimelineItems for the 「全部」 tab.

    Strategy: brute-query each table within (or near) the window, compute
    effective_at in app code, filter strictly, then merge-sort. For demo
    scale this is fine; for large data, push effective_at into SQL via
    GENERATED columns or CTE.
    """
    kinds = kinds or {"asset", "event", "contact", "input_turn", "file"}
    items: list = []

    # ── assets (joined to skill name) ──
    if "asset" in kinds:
        stmt = (
            select(Asset, GlobalSkill.name.label("skill_name"), UserSkill.render_spec, UserSkill.display_name)
            .join(UserSkill, Asset.user_skill_id == UserSkill.id)
            .join(GlobalSkill, UserSkill.skill_id == GlobalSkill.id)
            .where(Asset.user_id == user_id)
        )
        if skill_names:
            stmt = stmt.where(GlobalSkill.name.in_(skill_names))
        rows = (await db.execute(stmt)).all()
        for asset, skill_name, render_spec, display_name in rows:
            items.append(_asset_item(asset, skill_name, render_spec, display_name))

    # ── events ──
    if "event" in kinds:
        stmt = select(Event).where(Event.user_id == user_id)
        events = (await db.execute(stmt)).scalars().all()
        for ev in events:
            items.append(_event_item(ev))

    # ── contacts (first-class entity in its own table) ──
    if "contact" in kinds:
        contacts = (await db.execute(
            select(Contact).where(Contact.user_id == user_id)
        )).scalars().all()
        for c in contacts:
            items.append(_contact_item(c))

    # ── input_turns ──
    # Filter rule: typed inputs are "AI conversation history", not "life
    # events" → exclude from timeline. The DERIVED assets (todo / event /
    # ...) still appear, they're the real records. Voice and (future)
    # imported turns stay on timeline because they represent captured
    # moments in the user's life.
    if "input_turn" in kinds:
        stmt = select(InputTurn).where(
            InputTurn.user_id == user_id,
            InputTurn.source != "typed",
        )
        turns = (await db.execute(stmt)).scalars().all()
        breakdown = await _derived_breakdown(db, user_id, [str(t.id) for t in turns])
        for t in turns:
            items.append(_input_turn_item(t, breakdown.get(str(t.id))))

    # ── files ──
    if "file" in kinds:
        stmt = select(File).where(File.user_id == user_id)
        files = (await db.execute(stmt)).scalars().all()
        for f in files:
            items.append(_file_item(f))

    # ── window filter ──
    if from_dt or to_dt:
        def in_window(it):
            ea = _parse_iso(it.get("effective_at"))
            if ea is None:
                return False
            if from_dt and ea < from_dt:
                return False
            if to_dt and ea > to_dt:
                return False
            return True
        items = [it for it in items if in_window(it)]

    # ── sort: effective_at desc(newest first); tie-break by created_at desc ──
    def sort_key(it):
        ea = _parse_iso(it.get("effective_at")) or _EPOCH_MIN
        ca = _parse_iso(it.get("created_at"))   or _EPOCH_MIN
        return (ea, ca)
    items.sort(key=sort_key, reverse=True)

    return items[:limit]
